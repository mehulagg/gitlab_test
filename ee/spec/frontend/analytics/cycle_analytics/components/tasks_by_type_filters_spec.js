import { shallowMount, mount } from '@vue/test-utils';
import { GlNewDropdown, GlNewDropdownItem, GlSegmentedControl } from '@gitlab/ui';
import TasksByTypeFilters from 'ee/analytics/cycle_analytics/components/tasks_by_type_filters.vue';
import {
  TASKS_BY_TYPE_SUBJECT_ISSUE,
  TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST,
  TASKS_BY_TYPE_FILTERS,
} from 'ee/analytics/cycle_analytics/constants';
import { groupLabels } from '../mock_data';

const selectedLabelIds = [groupLabels[0].id];

const findSubjectFilters = ctx =>
  ctx.find('.js-tasks-by-type-chart-filters-subject').find(GlSegmentedControl);
const findSelectedSubjectFilters = ctx => findSubjectFilters(ctx).attributes('checked');
const findDropdown = ctx => ctx.find(GlNewDropdown);
const findDropdownLabels = ctx =>
  ctx.find('.js-tasks-by-type-chart-filters-labels').findAll(GlNewDropdownItem);

const selectLabelAtIndex = (ctx, index) => {
  findDropdownLabels(ctx)
    .at(index)
    .vm.$emit('click');
  return ctx.vm.$nextTick();
};

function createComponent({ props = {}, shallow = true }) {
  const fn = shallow ? shallowMount : mount;
  return fn(TasksByTypeFilters, {
    propsData: {
      selectedLabelIds,
      labels: groupLabels,
      subjectFilter: TASKS_BY_TYPE_SUBJECT_ISSUE,
      ...props,
    },
    stubs: {
      GlNewDropdown: true,
      GlNewDropdownItem: true,
    },
  });
}

describe('TasksByTypeFilters', () => {
  let wrapper = null;

  describe('with data available', () => {
    beforeEach(() => {
      wrapper = createComponent({});
    });

    afterEach(() => {
      wrapper.destroy();
    });

    it('should render the filters', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    describe('labels', () => {
      it(`should have ${selectedLabelIds.length} selected`, () => {
        expect(wrapper.text()).toContain('1 selected (15 max)');
      });

      describe('with label dropdown open', () => {
        beforeEach(() => {
          wrapper = createComponent({});
        });

        it('renders the group labels as dropdown items', () => {
          expect(findDropdown(wrapper).html()).toMatchSnapshot();
        });

        it('emits the `updateFilter` event when a subject label is clicked', () => {
          expect(wrapper.emitted().updateFilter).toBeUndefined();
          return selectLabelAtIndex(wrapper, 0).then(() => {
            expect(wrapper.emitted().updateFilter).toBeDefined();

            expect(wrapper.emitted().updateFilter[0]).toEqual([
              { filter: TASKS_BY_TYPE_FILTERS.LABEL, value: groupLabels[0].id },
            ]);
          });
        });

        describe('with maximum labels selected', () => {
          beforeEach(() => {
            wrapper = createComponent({
              props: {
                maxLabels: 2,
                selectedLabelIds: [groupLabels[0].id, groupLabels[1].id],
              },
            });
          });

          it('should not allow selecting another label', () => {
            expect(wrapper.emitted().updateFilter).toBeUndefined();

            return selectLabelAtIndex(wrapper, 2).then(() => {
              expect(wrapper.emitted().updateFilter).toBeUndefined();
            });
          });
        });
      });
    });

    describe('subject', () => {
      it('has subject filters', () => {
        expect(findSubjectFilters(wrapper).html()).toMatchSnapshot();
      });

      it('has the issue subject set by default', () => {
        expect(findSelectedSubjectFilters(wrapper)).toBe(TASKS_BY_TYPE_SUBJECT_ISSUE);
      });

      it('emits the `updateFilter` event when a subject filter is clicked', () => {
        wrapper = createComponent({ shallow: false });
        expect(wrapper.emitted().updateFilter).toBeUndefined();

        findSubjectFilters(wrapper)
          .findAll('label:not(.active)')
          .at(0)
          .trigger('click');

        return wrapper.vm.$nextTick(() => {
          expect(wrapper.emitted().updateFilter).toBeDefined();
          expect(wrapper.emitted().updateFilter[0]).toEqual([
            { filter: TASKS_BY_TYPE_FILTERS.SUBJECT, value: TASKS_BY_TYPE_SUBJECT_MERGE_REQUEST },
          ]);
        });
      });
    });
  });

  describe('no data available', () => {
    beforeEach(() => {
      wrapper = createComponent({});
    });

    afterEach(() => {
      wrapper.destroy();
    });

    it('should render the no data available message', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
