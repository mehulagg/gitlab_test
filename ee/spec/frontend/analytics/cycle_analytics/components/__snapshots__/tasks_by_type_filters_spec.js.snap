// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksByTypeFilters no data available should render the no data available message 1`] = `
"<div class=\\"js-tasks-by-type-chart-filters d-flex flex-row justify-content-between align-items-center\\">
  <div class=\\"flex-column\\">
    <h4>Tasks by type</h4>
    <p>Showing Issues and 1 labels</p>
  </div>
  <div class=\\"flex-column\\">
    <glnewdropdown-stub headertext=\\"\\" text=\\"\\" category=\\"tertiary\\" variant=\\"default\\" size=\\"medium\\" icon=\\"settings\\" aria-expanded=\\"false\\" aria-label=\\"CycleAnalytics|Display chart filters\\" right=\\"\\">
      <div class=\\"js-tasks-by-type-chart-filters-subject mb-3 px-3\\">
        <p class=\\"font-weight-bold text-left mb-2\\">Show</p>
        <gl-segmented-control-stub checked=\\"Issue\\" options=\\"[object Object],[object Object]\\"></gl-segmented-control-stub>
      </div>
      <gl-dropdown-divider-stub></gl-dropdown-divider-stub>
      <div class=\\"js-tasks-by-type-chart-filters-labels mb-3 px-3\\">
        <p class=\\"font-weight-bold text-left my-2\\">
          Select labels
          <br> <small>1 selected (2 max)</small></p>
        <gl-search-box-by-type-stub value=\\"\\" class=\\"js-tasks-by-type-chart-filters-subject mb-2\\"></gl-search-box-by-type-stub>
        <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" ischecked=\\"true\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 0, 0);\\"></span>
          roses
        </glnewdropdownitem-stub>
        <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 255, 255);\\"></span>
          some space
        </glnewdropdownitem-stub>
        <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(0, 0, 255);\\"></span>
          violets
        </glnewdropdownitem-stub>
        <div class=\\"text-secondary\\" style=\\"display: none;\\">
          No matching labels
        </div>
      </div>
    </glnewdropdown-stub>
  </div>
</div>"
`;

exports[`TasksByTypeFilters with data available labels with label dropdown open renders the group labels as dropdown items 1`] = `
"<glnewdropdown-stub headertext=\\"\\" text=\\"\\" category=\\"tertiary\\" variant=\\"default\\" size=\\"medium\\" icon=\\"settings\\" aria-expanded=\\"false\\" aria-label=\\"CycleAnalytics|Display chart filters\\" right=\\"\\">
  <div class=\\"js-tasks-by-type-chart-filters-subject mb-3 px-3\\">
    <p class=\\"font-weight-bold text-left mb-2\\">Show</p>
    <gl-segmented-control-stub checked=\\"Issue\\" options=\\"[object Object],[object Object]\\"></gl-segmented-control-stub>
  </div>
  <gl-dropdown-divider-stub></gl-dropdown-divider-stub>
  <div class=\\"js-tasks-by-type-chart-filters-labels mb-3 px-3\\">
    <p class=\\"font-weight-bold text-left my-2\\">
      Select labels
      <br> <small>1 selected (2 max)</small></p>
    <gl-search-box-by-type-stub value=\\"\\" class=\\"js-tasks-by-type-chart-filters-subject mb-2\\"></gl-search-box-by-type-stub>
    <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" ischecked=\\"true\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 0, 0);\\"></span>
      roses
    </glnewdropdownitem-stub>
    <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 255, 255);\\"></span>
      some space
    </glnewdropdownitem-stub>
    <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(0, 0, 255);\\"></span>
      violets
    </glnewdropdownitem-stub>
    <div class=\\"text-secondary\\" style=\\"display: none;\\">
      No matching labels
    </div>
  </div>
</glnewdropdown-stub>"
`;

exports[`TasksByTypeFilters with data available should render the filters 1`] = `
"<div class=\\"js-tasks-by-type-chart-filters d-flex flex-row justify-content-between align-items-center\\">
  <div class=\\"flex-column\\">
    <h4>Tasks by type</h4>
    <p>Showing Issues and 1 labels</p>
  </div>
  <div class=\\"flex-column\\">
    <glnewdropdown-stub headertext=\\"\\" text=\\"\\" category=\\"tertiary\\" variant=\\"default\\" size=\\"medium\\" icon=\\"settings\\" aria-expanded=\\"false\\" aria-label=\\"CycleAnalytics|Display chart filters\\" right=\\"\\">
      <div class=\\"js-tasks-by-type-chart-filters-subject mb-3 px-3\\">
        <p class=\\"font-weight-bold text-left mb-2\\">Show</p>
        <gl-segmented-control-stub checked=\\"Issue\\" options=\\"[object Object],[object Object]\\"></gl-segmented-control-stub>
      </div>
      <gl-dropdown-divider-stub></gl-dropdown-divider-stub>
      <div class=\\"js-tasks-by-type-chart-filters-labels mb-3 px-3\\">
        <p class=\\"font-weight-bold text-left my-2\\">
          Select labels
          <br> <small>1 selected (2 max)</small></p>
        <gl-search-box-by-type-stub value=\\"\\" class=\\"js-tasks-by-type-chart-filters-subject mb-2\\"></gl-search-box-by-type-stub>
        <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" ischecked=\\"true\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 0, 0);\\"></span>
          roses
        </glnewdropdownitem-stub>
        <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(255, 255, 255);\\"></span>
          some space
        </glnewdropdownitem-stub>
        <glnewdropdownitem-stub avatarurl=\\"\\" iconcolor=\\"\\" iconname=\\"\\" iconrightname=\\"\\" secondarytext=\\"\\"><span class=\\"d-inline-block dropdown-label-box\\" style=\\"background-color: rgb(0, 0, 255);\\"></span>
          violets
        </glnewdropdownitem-stub>
        <div class=\\"text-secondary\\" style=\\"display: none;\\">
          No matching labels
        </div>
      </div>
    </glnewdropdown-stub>
  </div>
</div>"
`;

exports[`TasksByTypeFilters with data available subject has subject filters 1`] = `"<gl-segmented-control-stub checked=\\"Issue\\" options=\\"[object Object],[object Object]\\"></gl-segmented-control-stub>"`;
